### PR TITLE
Add Module Loader runtime objects

### PR DESCRIPTION
**Pull request by boingoing**
Based on the W3C Module Loader spec:
https://whatwg.github.io/loader

Includes more or less complete implementations of the runtime objects:
%Registry%
%RegistryPrototype%
%Loader%
%LoaderPrototype%
%ModuleStatus%
%ModuleStatusPrototype%

Includes a stub implementation of %Module%.

Most of the abstract operations in the spec are implemented completely but
we are missing some key ones such as SatisfyInstance. The next step will
be to merge the existing ModuleRecord implementation with the ModuleStatus
object and set of abstract operations.

In particular, this pull request does not plug-in any of the dependency
graph resolution or export name resolution because a refactor of existing
ModuleRecord code is required.

There are some open issues with the code in this change. For example, it's
not clear from the spec what we should do if transforming a promise object
with a fulfillment handler returns a non-promise object. It seems possible
that user code could replace the constructor property on the promise
instance which could cause the transformed promise to be anything. Should
we throw here? Should we just wrap whatever the object is in a resolved
promise? It's not clear. Most of the machinery in the spec requires
promise objects so I have taken the approach that we should throw anytime
an object is not a promise when we expect it should be.

### CODE REVIEW

**Comment by boingoing** [comment]
@Yongqu Please take a look over this one, thanks! 😀

File: lib/Runtime/Library/JavascriptBuiltInFunctionList.h [hunk 1]
```diff
@@ -962,5 +962,39 @@ BUILTIN(JavascriptGeneratorFunction, NewInstance, NewInstance, FunctionInfo::Ski
...
+BUILTIN(JavascriptRegistry, Keys, EntryKeys, FunctionInfo::ErrorOnNew)
+BUILTIN(JavascriptRegistry, Values, EntryValues, FunctionInfo::ErrorOnNew)
+BUILTIN(JavascriptLoader, NewInstance, NewInstance, FunctionInfo::SkipDefaultNewObject)
+BUILTIN(JavascriptLoader, Import, EntryImport, FunctionInfo::ErrorOnNew)
+BUILTIN(JavascriptLoader, Resolve, EntryResolve, FunctionInfo::ErrorOnNew)
+BUILTIN(JavascriptLoader, Load, EntryLoad, FunctionInfo::ErrorOnNew)
+BUILTIN(JavascriptLoader, GetterRegistry, EntryGetterRegistry, FunctionInfo::ErrorOnNew)
+BUILTIN(JavascriptLoader, EnsureRegisteredAndEvaluatedFulfillmentHandler, EntryEnsureRegisteredAndEvaluatedFulfillmentHandler, FunctionInfo::ErrorOnNew | FunctionInfo::DoNotProfile)
+BUILTIN(JavascriptLoader, EnsureRegisteredFulfillmentHandler, EntryEnsureRegisteredFulfillmentHandler, FunctionInfo::ErrorOnNew | FunctionInfo::DoNotProfile)
+BUILTIN(JavascriptLoader, EnsureEvaluatedFulfillmentHandler, EntryEnsureEvaluatedFulfillmentHandler, FunctionInfo::ErrorOnNew | FunctionInfo::DoNotProfile)
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptBuiltInFunctionList.h]: is this an official builtin?

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 2]
```diff
@@ -1357,6 +1403,26 @@ namespace Js
             DeferredTypeHandler<InitializeWeakSetConstructor>::GetDefaultInstance());
         AddFunction(globalObject, PropertyIds::WeakSet, weakSetConstructor);
 
+        registryConstructor = nullptr;
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: nitpick: either we all do it pre-initialization, or all do it if () else() style. seems like we are more consistent in doing if...else style

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 3]
```diff
@@ -1357,6 +1403,26 @@ namespace Js
...
+                DeferredTypeHandler<InitializeRegistryConstructor>::GetDefaultInstance());
+
+            loaderConstructor = CreateBuiltinConstructor(&JavascriptLoader::EntryInfo::NewInstance,
+                DeferredTypeHandler<InitializeLoaderConstructor>::GetDefaultInstance());
+
+            moduleStatusConstructor = CreateBuiltinConstructor(&JavascriptModuleStatus::EntryInfo::NewInstance,
+                DeferredTypeHandler<InitializeModuleStatusConstructor>::GetDefaultInstance());
+
+            // TODO: This should construct something else
+            moduleConstructor = CreateBuiltinConstructor(&JavascriptModuleStatus::EntryInfo::NewInstance,
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: set to nullptr for now to avoid confusion?

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 4]
```diff
@@ -2178,6 +2244,147 @@ namespace Js
...
+
+        library->AddMember(loaderConstructor, PropertyIds::length, TaggedInt::ToVarUnchecked(0), PropertyConfigurable);
+        library->AddMember(loaderConstructor, PropertyIds::prototype, library->loaderPrototype, PropertyNone);
+
+        if (scriptContext->GetConfig()->IsES6FunctionNameEnabled())
+        {
+            library->AddMember(loaderConstructor, PropertyIds::name, library->CreateStringFromCppLiteral(_u("Loader")), PropertyConfigurable);
+        }
+
+        library->AddMember(loaderConstructor, PropertyIds::resolve, library->GetSymbolResolve(), PropertyNone);
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: why these 4?

File: lib/Runtime/Types/EdgeJavascriptTypeId.h [hunk 5]
```diff
@@ -110,17 +110,20 @@ enum TypeId
     TypeIds_JavascriptEnumeratorIterator = 67,
     TypeIds_Generator = 68,
     TypeIds_Promise = 69,
+    TypeIds_Registry = 70,
+    TypeIds_Loader = 71,
```
Comment (Yongqu) [inline lib/Runtime/Types/EdgeJavascriptTypeId.h]: ModuleLoader/ModuleRegistry to clarify?

File: lib/Runtime/Library/JavascriptModuleStatus.cpp [hunk 6]
```diff
@@ -0,0 +1,1282 @@
...
+
+        Var fetchVar = JavascriptOperators::GetProperty(entry->loader, Js::PropertyIds::_symbolFetch, scriptContext);
+        if (!JavascriptConversion::IsCallable(fetchVar))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedFunction);
+        }
+
+        RecyclableObject* hook = RecyclableObject::FromVar(fetchVar);
+
+        Var hookResult = CALL_FUNCTION(hook, Js::CallInfo(CallFlags_Value, 3),
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptModuleStatus.cpp]: we should probe stack at the beginning if we use CALL_FUNCTION here.

File: lib/Runtime/Library/JavascriptModuleStatus.cpp [hunk 7]
```diff
@@ -0,0 +1,1282 @@
...
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedPromise);
+        }
+
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+        JavascriptFunction* throwerFunction = library->GetThrowerFunction();
+        JavascriptModuleStatusFulfillmentHandlerFunction* fulfillmentHandler = library->CreateModuleStatusFulfillmentHandlerFunction(
+            EntryUpgradeToStageFulfillmentHandler, 
+            &JavascriptModuleStatus::EntryInfo::UpgradeToStageFulfillmentHandler, 
+            entry, 
+            JavascriptModuleStatusStage::Translate,
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptModuleStatus.cpp]: we need to pass the "payload/source" around. I think more about that

File: lib/Runtime/Library/JavascriptModuleStatus.cpp [hunk 8]
```diff
@@ -0,0 +1,1282 @@
...
+            return JavascriptPromise::CreateRejectedPromise(rangeError, scriptContext);
+            }
+        }
+
+        Assert(requestResult != nullptr);
+
+        return JavascriptPromise::CreatePassThroughPromise(requestResult, scriptContext);
+    }
+
+    JavascriptPromise* JavascriptModuleStatus::RequestFetch(JavascriptModuleStatus* entry, ScriptContext* scriptContext)
```
Comment (Yongqu) [inline lib/Runtime/Library/JavascriptModuleStatus.cpp]: pretty cool :)

**Comment by Yongqu** [comment]
:shipit:

**Comment by dilijev** [comment]
Followed up with @boingoing offline. Seems some design change might render this PR obsolete but salvageable. Adding Waiting label.

**Comment by boingoing** [comment]
Closing this for now, it doesn't appear that we will be implementing this in the near future.

**Pull request closed without merging**